pennylane/data/attributes/sparse_array.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains AttributeType definition for ``scipy.sparse.csr_array``."""
+
+from typing import Generic, Tuple, Type, TypeVar, Union, cast
+
+import numpy as np
+from scipy.sparse import (
+    bsr_array,
+    bsr_matrix,
+    coo_array,
+    coo_matrix,
+    csc_array,
+    csc_matrix,
+    csr_array,
+    csr_matrix,
+    dia_array,
+    dia_matrix,
+    dok_array,
+    dok_matrix,
+    lil_array,
+    lil_matrix,
+)
+
+from pennylane.data.base.attribute import AttributeInfo, AttributeType
+from pennylane.data.base.typing_util import HDF5Group
+
+_ALL_SPARSE = (
+    bsr_array,
+    coo_array,
+    csc_array,
+    csr_array,
+    dia_array,
+    dok_array,
+    lil_array,
+    csc_matrix,
+    csr_matrix,
+    bsr_matrix,
+    coo_matrix,
+    dia_matrix,
+    dok_matrix,
+    lil_matrix,
+)
+_ALL_SPARSE_MAP = {type_.__qualname__: type_ for type_ in _ALL_SPARSE}
+
+SparseArray = Union[bsr_array, coo_array, csc_array, csr_array, dia_array, dok_array, lil_array]
+SparseArrayT = TypeVar("SparseArrayT", bound=SparseArray)
+
+
+class DatasetSparseArray(
+    Generic[SparseArrayT], AttributeType[HDF5Group, SparseArrayT, SparseArrayT]
+):
+    """Attribute type for Scipy sparse arrays. Can accept values of any type in
+    ``scipy.sparse``. Arrays are stored in CSR format."""
```

Comment:
Is CSR an obvious choice? Should this be a flexible option?

Reply:
I chose CSR because that's what's used for existing datasets. I don't think the choice of serialized format should be flexible, since they differ in subtle ways.